contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,214 @@
+ * with L1 but instead must communicate with that L1 via an L2. Each contract that extends the ForwarderBase maintains a mapping of 
+ * contracts that exist on networks that are connected to the network that this contract is deployed on. For example, if this contract is deployed on
+ * Arbitrum, then there could be a mapping of contracts for each L3 that is connected to Arbitrum and communicates to L1 via Arbitrum. In other words,
+ * this contract maintains a mapping of important contracts on the "next layer" and is used to help relay messages from L1 to L3 by sending messages
+ * through the next layer. Messages could be forwarded across an indefinite amount of layers by recursively wrapping and sending messages over canonical
+ * bridges, only stopping when the following network contains the intended target contract.
+ * @custom:security-contact bugs@across.to
+ */
+abstract contract ForwarderBase is UUPSUpgradeable, AdapterInterface {
+    /**
+     * @notice Route contains all information needed in order to send a message to the next layer. It describes two possible paths: a message is sent
+     * through `adapter` (on this network) to the `forwarder` on the next layer, or a message is sent through `adapter` to the `spokePool`.
+     * @dev We need to know three different addresses:
+     *   - The address of the adapter to delegatecall, so that we may propagate message and token relays. This contract lives on this layer.
+     *   - The address of the next layer's forwarder, so that we can continue the message and token relay chain.
+     *   - The address of the next layer's spoke pool, so that we can determine whether the following network contains the target spoke pool.
+     * these addresses describe the path a message or token relay must take to arrive at the next layer. A Route only describes the path to arrive
+     * a layer closer to to the ultimate target, unless one of the `forwarder` or `spokePool` addresses equal the target contract, in which case
+     * the path ends. All Routes must contain an adapter address; however, not all Routes must contain a forwarder or spokePool address. For example,
+     * we may send a message to a spoke pool on a network with no corresponding forwarder contract, or we may send a message to a network which contains
+     * a forwarder contract but no spoke pool.
+     */
+    struct Route {
+        address adapter;
+        address forwarder;
+        address spokePool;
+    }
+
+    // Map from a destination address to the next token/message bridge route to reach the destination address. The destination address can either be a
+    // spoke pool or a forwarder. Destination addresses are used as the key in this mapping because the `target` address is the only information
+    // we receive when `relay{Message/Token}` is called on this contract, so we assume that these destination contracts on all the possible next
```

Comment:
I don't really see why we can't just include the chainId? The only thing we'd need is that each chain has a unique adapter on mainnet, right?

Reply:
do you mean add `chainId` to the `Route` struct?

Reply:
The issue with using the chain id was `relayTokens`. Ideally, we send tokens to the target contract and then call `relayMessage`, which calls `relayTokens` on the forwarder, but since `relayTokens` (as we define it in `AdapterInterface`) has no way of adding extra data, we would probably have to make a special `relayTokens` function and therefore a new interface. Far from untenable, but it was simpler to use the addresses instead.

Reply:
> Far from untenable

How would we do it? Just curious, maybe its worth doing since its a much more natural key than `destinationAddress`

Reply:
(Perhaps far from optimal) but a simple way is if the chain id is stored on the L1 adapter, then we just send a message to the forwarder which also includes this chain id (like `abi.encode(chainId, message)`) which can be processed by all subsequent forwarders. If we made a new relayTokens function, it could look something like `relayTokens(address baseToken, address remoteToken, uint256 to, address to, uint256 destinationChainId)` and then for each forwarder keep encoding this message on every token relay.

Reply:
> a simple way is if the chain id is stored on the L1 adapter, then we just send a message to the forwarder which also includes this chain id (like `abi.encode(chainId, message)`) which can be processed by all subsequent forwarders.

This makes sense to me. The router contract on the L1 would need to know this (set in the constructor arguments) and then each subsequent recipient would keep propagating that same message until the next-hop bridge is owned by that destination chain ID. I think it could simplify the logic quite a bit.

Reply:
Done here https://github.com/across-protocol/contracts/pull/609/commits/41b4856a3bf93ba93296db1a38ea69b899e9f926